x/accounts/vesting/continuous_vesting_account.go

```diff
@@ -0,0 +1,167 @@
+
+	// We must handle the case where the start time for a vesting account has
+	// been set into the future or when the start of the chain is not exactly
+	// known.
+	startTime, err := cva.StartTime.Get(ctx)
+	if err != nil {
+		return nil, nil, err
+	}
+	endTime, err := cva.EndTime.Get(ctx)
+	if err != nil {
+		return nil, nil, err
+	}
+	var originalVesting sdk.Coins
+	cva.IterateCoinEntries(ctx, cva.OriginalVesting, func(key string, value math.Int) (stop bool) {
+		originalVesting = append(originalVesting, sdk.NewCoin(key, value))
+		return false
+	})
+	if startTime.GTE(math.NewInt(blockTime.Unix())) {
+		return vestedCoins, originalVesting, nil
+	} else if endTime.LTE(math.NewInt(blockTime.Unix())) {
+		return originalVesting, vestingCoins, nil
+	}
+
+	// calculate the vesting scalar
+	x := math.NewInt(blockTime.Unix()).Sub(startTime).Int64()
+	y := endTime.Sub(startTime).Int64()
+	s := math.LegacyNewDec(x).Quo(math.LegacyNewDec(y))
+
+	for _, ovc := range originalVesting {
+		vestedAmt := math.LegacyNewDecFromInt(ovc.Amount).Mul(s).RoundInt()
+		vestedCoins = append(vestedCoins, sdk.NewCoin(ovc.Denom, vestedAmt))
+	}
+
+	vestingCoins = originalVesting.Sub(vestedCoins...)
+
+	return vestedCoins, vestingCoins, nil
+}
+
+// GetVestingCoins returns the total number of vesting coins. If no coins are
+// vesting, nil is returned.
+func (cva ContinuousVestingAccount) GetVestingCoins(ctx context.Context, blockTime time.Time) (sdk.Coins, error) {
+	_, vestingCoins, err := cva.GetVestCoinsInfo(ctx, blockTime)
+	if err != nil {
+		return nil, err
+	}
+	return vestingCoins, nil
+}
+
+func (cva ContinuousVestingAccount) QueryVestCoinsInfo(ctx context.Context, msg *vestingtypes.QueryVestCoinsInfoRequest) (
+	*vestingtypes.QueryVestCoinsInfoResponse, error,
+) {
+	hs := cva.headerService.GetHeaderInfo(ctx)
+	vestedCoins, vestingCoins, err := cva.GetVestCoinsInfo(ctx, hs.Time)
+	if err != nil {
+		return nil, err
+	}
+	return &vestingtypes.QueryVestCoinsInfoResponse{
+		VestedVesting: vestedCoins,
+		VestingCoins:  vestingCoins,
+	}, nil
+}
+
+func (cva ContinuousVestingAccount) QueryStartTime(ctx context.Context, msg *vestingtypes.QueryStartTimeRequest) (
+	*vestingtypes.QueryStartTimeResponse, error,
+) {
+	startTime, err := cva.StartTime.Get(ctx)
+	if err != nil {
+		return nil, err
+	}
+	return &vestingtypes.QueryStartTimeResponse{
+		StartTime: startTime.Int64(),
+	}, nil
+}
+
+// Implement smart account interface
+func (cva ContinuousVestingAccount) RegisterInitHandler(builder *accountstd.InitBuilder) {
+	accountstd.RegisterInitHandler(builder, cva.Init)
+}
+
+func (cva ContinuousVestingAccount) RegisterExecuteHandlers(builder *accountstd.ExecuteBuilder) {
+	accountstd.RegisterExecuteHandler(builder, cva.ExecuteMessages)
+	cva.BaseVesting.RegisterExecuteHandlers(builder)
+}
+
+func (cva ContinuousVestingAccount) RegisterQueryHandlers(builder *accountstd.QueryBuilder) {
+	accountstd.RegisterQueryHandler(builder, cva.QueryStartTime)
```

Comment:
I would group the auxiliary queries and the base account queries into one query, having one query just for start time and vest info seems a little overkill, and most likely clients will want to fetch all the info in one go so I'd advocate for a single query